### ISSUE TITLE
c++ api; scatter_sum works on kCPU, but not kCUDA

### ISSUE THREAD

**Issue by JellePiepenbrock** [rusty1s/pytorch_scatter#326]
I'm trying to implement a scatter_sum operation via the c++ api.

I'm calling the function as follows:

`results = scatter_sum(source_nodes, target_index_tensor, dim, torch::nullopt, torch::nullopt);`

I have verified that both tensors are on cuda:0 via these lines:
```
std::cout << source_nodes.device() << std::endl;
std::cout << target_index_tensor.device() << std::endl;
```
The program simply fails when I used 'kCUDA' as the device, but when I use 'kCPU' as the device, it works. I have verified that the normal torch functions (linear, relu) work on the kCUDA device, so only this scatter_sum function does not go through. What could be the cause of the program failing? I simply get 'core dumped', but because it works on CPU, it's not so clear to me what could be wrong.

Some information about the system:
Python 3.8
CUDA 10.2
PyTorch 1.10

**Comment by rusty1s** [comment]
Sorry for late reply. Do other versions work for you such as `scatter_max`? For `scatter_add`, you can also directly call the PyTorch equivalent: https://pytorch.org/cppdocs/api/function_namespaceat_1af0efb124efa03c8efb8d8bba637ab296.html?highlight=scatter_add

**Comment by JellePiepenbrock** [comment]
So the previous issue was resolved (I didn't have the correct CMAKE settings). Now, everything compiles without a problem and there are files in the cuda folder etc. 

However, I'm running into a new (for me) issue that can be described in a very similar way (which is why I'm not making a new issue):

When I use scatter_sum or scatter_max on kCPU, it works and the GNN I'm training converges to reasonable solutions. However, when I use kCUDA as the device, scatter_sum gives me only 0-vectors as output (regardless of input). Does this sound like a problem you know about?

**Comment by JellePiepenbrock** [comment]
There might still be something wrong with the compilation; if I use scatter_sum, compilation works but I get 0s as output (on kCUDA, on kCPU it works as expected). However when I try to use scatter_min or scatter_max, the compilation fails;

`error: no match for ‘operator=’ (operand types are ‘at::Tensor’ and ‘std::tuple<at::Tensor, at::Tensor>’)
  525 |       results = scatter_max(source_nodes, target_index_tensor, dim, torch::nullopt, torch::nullopt);`

**Comment by rusty1s** [comment]
Mh, what happens if you install the Python version of these using `python setup.py install`? Does `scatter_max` work as expected in this case?

**Comment by JellePiepenbrock** [comment]
Okay, thank you for the suggestion. The python script gave a very clear error that indicates there is a CUDA version conflict. When I fix the version number for cudatoolkit (cudatoolkit=10.2) via conda, the python version works.

However, it seems that the version of cudatoolkit installed via conda is not enough to compile the C++ version on its own (it says no cuda compiler (nvcc) can be found (and I also don't see it in env's lib folder). Do you have any experience or workflow that solves this (preferably applicable to a conda-type environment)? There is a conda package cuda-toolkit-dev that seems to include nvcc etc. , but this is unfortunately missing exactly the version (10.2) I need.

**Comment by rusty1s** [comment]
I think there is a `conda-forge` package called `cudatoolkit-dev` which should include `nvcc`.

**Comment by rusty1s** [comment]
Yeah, my suggestion was also to take a look at `cudatoolkit-dev`. For CUDA 10.2, you might need to indeed download the file from NVIDIA directly :(

https://developer.nvidia.com/cuda-10.2-download-archive
